File: chaos-flutter/lib/features/widgets/room_card.dart
import 'dart:io';

import 'package:fluent_ui/fluent_ui.dart' as fluent;
import 'package:flutter/material.dart';

import '../../core/models/live_directory.dart';
import 'network_image_view.dart';

String? formatOnlineCount(int? n) {
  if (n == null) return null;
  if (n < 0) return null;
  if (n < 10000) return n.toString();
  final w = (n / 10000.0);
  return '${w.toStringAsFixed(w >= 10 ? 0 : 1)}万';
}

class RoomCard extends StatelessWidget {
  const RoomCard({
    super.key,
    required this.room,
    this.onTap,
  });

  final LiveDirRoomCard room;
  final VoidCallback? onTap;

  @override
  Widget build(BuildContext context) {
    return Platform.isWindows
        ? _WindowsRoomCard(room: room, onTap: onTap)
        : _MaterialRoomCard(room: room, onTap: onTap);
  }
}

class _MaterialRoomCard extends StatelessWidget {
  const _MaterialRoomCard({required this.room, this.onTap});
  final LiveDirRoomCard room;
  final VoidCallback? onTap;

  @override
  Widget build(BuildContext context) {
    final online = formatOnlineCount(room.online);
    final cs = Theme.of(context).colorScheme;

    return LayoutBuilder(
      builder: (context, c) {
        // 卡片高度较小时（例如网格列数很多时），降低信息密度避免 RenderFlex overflow。
        final compact = c.maxHeight > 0 && c.maxHeight < 180;
        final titleLines = compact ? 1 : 2;
        final showUser = !compact;
        final pad = compact
            ? const EdgeInsets.fromLTRB(12, 8, 12, 10)
            : const EdgeInsets.fromLTRB(12, 10, 12, 12);

        return Card(
          clipBehavior: Clip.antiAlias,
          child: InkWell(
            onTap: onTap,
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                AspectRatio(
                  aspectRatio: 16 / 9,
                  child: Stack(
                    fit: StackFit.expand,
                    children: [
                      NetworkImageView(url: room.cover, borderRadius: 0),
                      if (online != null)
                        Positioned(
                          left: 0,
                          right: 0,
                          bottom: 0,
                          child: Container(
                            decoration: BoxDecoration(
                              gradient: LinearGradient(
                                begin: Alignment.bottomCenter,
                                end: Alignment.topCenter,
                                colors: [
                                  Colors.black.withOpacity(0.75),
                                  Colors.transparent,
                                ],
                              ),
                            ),
                            padding: const EdgeInsets.fromLTRB(6, 10, 6, 6),
                            child: Row(
                              mainAxisAlignment: MainAxisAlignment.end,
                              children: [
                                const Icon(Icons.local_fire_department,
                                    color: Colors.white, size: 14),
                                const SizedBox(width: 4),
                                Text(
                                  online,
                                  style: const TextStyle(
                                    fontSize: 12,
                                    color: Colors.white,
                                    fontWeight: FontWeight.w600,
                                  ),
                                ),
                              ],
                            ),
                          ),
                        ),
                    ],
                  ),
                ),
                Padding(
                  padding: pad,
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(
                        room.title,
                        maxLines: titleLines,
                        overflow: TextOverflow.ellipsis,
                        style: Theme.of(context)
                            .textTheme
                            .titleSmall
                            ?.copyWith(fontWeight: FontWeight.w600),
                      ),
                      if (showUser) ...[
                        const SizedBox(height: 6),
                        Text(
                          room.userName ?? '',
                          maxLines: 1,
                          overflow: TextOverflow.ellipsis,
                          style: Theme.of(context)
                              .textTheme
                              .bodySmall
                              ?.copyWith(color: cs.onSurfaceVariant),
                        ),
                      ],
                    ],
                  ),
                ),
              ],
            ),
          ),
        );
      },
    );
  }
}

class _WindowsRoomCard extends StatelessWidget {
  const _WindowsRoomCard({required this.room, this.onTap});
  final LiveDirRoomCard room;
  final VoidCallback? onTap;

  @override
  Widget build(BuildContext context) {
    final online = formatOnlineCount(room.online);

    return LayoutBuilder(
      builder: (context, c) {
        final compact = c.maxHeight > 0 && c.maxHeight < 180;
        final titleLines = compact ? 1 : 2;
        final showUser = !compact;
        final pad = compact
            ? const EdgeInsets.fromLTRB(12, 8, 12, 10)
            : const EdgeInsets.fromLTRB(12, 10, 12, 12);

        // fluent_ui 的 Card + Button 组合，视觉上更接近 WinUI3 的“卡片”交互。
        return fluent.Button(
          onPressed: onTap,
          style: const fluent.ButtonStyle(
            padding: fluent.WidgetStatePropertyAll(EdgeInsets.zero),
          ),
          child: fluent.Card(
            padding: EdgeInsets.zero,
            child: fluent.Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                fluent.AspectRatio(
                  aspectRatio: 16 / 9,
                  child: Stack(
                    fit: StackFit.expand,
                    children: [
                      // Windows 端直接用 Image.network，避免额外依赖。
                      ClipRRect(
                        borderRadius: const BorderRadius.only(
                          topLeft: Radius.circular(8),
                          topRight: Radius.circular(8),
                        ),
                        child: (room.cover ?? '').trim().isEmpty
                            ? Container(color: Colors.black12)
                            : Image.network(
                                (room.cover ?? '').trim(),
                                fit: BoxFit.cover,
                                errorBuilder: (context, _, __) =>
                                    Container(color: Colors.black12),
                                loadingBuilder: (context, child, progress) {
                                  if (progress == null) return child;
                                  return Container(color: Colors.black12);
                                },
                              ),
                      ),
                      if (online != null)
                        Positioned(
                          left: 0,
                          right: 0,
                          bottom: 0,
                          child: Container(
                            decoration: BoxDecoration(
                              gradient: LinearGradient(
                                begin: Alignment.bottomCenter,
                                end: Alignment.topCenter,
                                colors: [
                                  Colors.black.withOpacity(0.75),
                                  Colors.transparent,
                                ],
                              ),
                            ),
                            padding: const EdgeInsets.fromLTRB(6, 10, 6, 6),
                            child: Row(
                              mainAxisAlignment: MainAxisAlignment.end,
                              children: [
                                const Icon(Icons.local_fire_department,
                                    color: Colors.white, size: 14),
                                const SizedBox(width: 4),
                                Text(
                                  online,
                                  style: const TextStyle(
                                    fontSize: 12,
                                    color: Colors.white,
                                    fontWeight: FontWeight.w600,
                                  ),
                                ),
                              ],
                            ),
                          ),
                        ),
                    ],
                  ),
                ),
                fluent.Padding(
                  padding: pad,
                  child: fluent.Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      fluent.Text(
                        room.title,
                        maxLines: titleLines,
                        overflow: TextOverflow.ellipsis,
                        style: const fluent.TextStyle(
                            fontSize: 13, fontWeight: FontWeight.w600),
                      ),
                      if (showUser) ...[
                        const fluent.SizedBox(height: 6),
                        fluent.Text(
                          room.userName ?? '',
                          maxLines: 1,
                          overflow: TextOverflow.ellipsis,
                          style: const fluent.TextStyle(fontSize: 12),
                        ),
                      ],
                    ],
                  ),
                ),
              ],
            ),
          ),
        );
      },
    );
  }
}

File: chaos-android/app/src/main/java/com/zerodevi1/chaos_seed/ui/components/RoomCard.kt
package com.zerodevi1.chaos_seed.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.LocalFireDepartment
import androidx.compose.material3.Card
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.SubcomposeLayout
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.zerodevi1.chaos_seed.core.model.LiveDirRoomCard

@Composable
fun RoomCard(
    room: LiveDirRoomCard,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Card(modifier = modifier.clickable(onClick = onClick)) {
        // Use a layout pass to decide whether we are "compact" (grid cell height too small).
        SubcomposeLayout { constraints ->
            val maxH = constraints.maxHeight
            val compact = maxH in 1..179
            val titleLines = if (compact) 1 else 2
            val showUser = !compact
            val pad = if (compact) PaddingValues(start = 12.dp, top = 8.dp, end = 12.dp, bottom = 10.dp)
            else PaddingValues(start = 12.dp, top = 10.dp, end = 12.dp, bottom = 12.dp)

            val p = subcompose("content") {
                Column {
                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .aspectRatio(16f / 9f),
                    ) {
                        NetworkImage(url = room.cover, modifier = Modifier.fillMaxSize())

                        val online = formatOnlineCount(room.online)
                        if (online != null) {
                            Box(
                                modifier = Modifier
                                    .align(Alignment.BottomCenter)
                                    .fillMaxWidth()
                                    .background(
                                        brush = Brush.verticalGradient(
                                            colors = listOf(
                                                Color.Black.copy(alpha = 0.75f),
                                                Color.Transparent,
                                            ),
                                        ),
                                    )
                                    .padding(start = 6.dp, top = 10.dp, end = 6.dp, bottom = 6.dp),
                            ) {
                                Row(
                                    horizontalArrangement = Arrangement.spacedBy(4.dp),
                                    verticalAlignment = Alignment.CenterVertically,
                                    modifier = Modifier.align(Alignment.CenterEnd),
                                ) {
                                    Icon(
                                        imageVector = Icons.Outlined.LocalFireDepartment,
                                        contentDescription = null,
                                        tint = Color.White,
                                    )
                                    Text(
                                        text = online,
                                        color = Color.White,
                                        style = MaterialTheme.typography.labelSmall.copy(fontWeight = FontWeight.SemiBold),
                                    )
                                }
                            }
                        }
                    }

                    Column(modifier = Modifier.padding(pad)) {
                        Text(
                            text = room.title,
                            maxLines = titleLines,
                            overflow = TextOverflow.Ellipsis,
                            style = MaterialTheme.typography.titleSmall.copy(fontWeight = FontWeight.SemiBold),
                        )
                        if (showUser) {
                            Text(
                                text = room.userName.orEmpty(),
                                maxLines = 1,
                                overflow = TextOverflow.Ellipsis,
                                style = MaterialTheme.typography.bodySmall.copy(color = MaterialTheme.colorScheme.onSurfaceVariant),
                                modifier = Modifier.padding(top = 6.dp),
                            )
                        }
                    }
                }
            }.map { it.measure(constraints) }

            val w = constraints.maxWidth
            val h = p.maxOfOrNull { it.height } ?: 0
            layout(w, h) { p.forEach { it.placeRelative(0, 0) } }
        }
    }
}
